enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,291 @@
+                command:
+                  - sleep
+                  - "3600"
+    strategy: deployment
+  installModes:
+  - supported: false
+    type: OwnNamespace
+  - supported: true
+    type: SingleNamespace
+  - supported: true
+    type: MultiNamespace
+  - supported: true
+    type: AllNamespaces
+  maturity: alpha
+  provider:
+    name: Red Hat
+  version: 1.0.0
+```
+
+The operator bundle should include the following Service:
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: foo-operator-readiness-service
+spec:
+  selector:
+    app: foo-operator
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+```
+
+#### CSVless Bundles
+
+In the case of a CSVless bundle, an annotation can be applied to the `annotations.yaml` which incdicates the name of the upgradeReadinessService and the deployment name:
+
+```yaml
+  operators.operatorframework.io.deployment.<DeploymentName>.readiness.service: "<ServiceName>"
+```
+
+Once the `annotations.yaml` is created with the key/value above, the following resources should appear in the operator bundle:
+
+- A service whose name matches the `ServiceName` in your annotation which routes to the readiness endpoint.
+- A deployment
+  - Whose name matches the DeploymentName used in your annotation
+  - Has a `podTemplate` with an annotation that is selected by your label
+
+### Risks and Mitigations
+
+- Not all operators may take advanatage of this feature. If an operator does not opt-in to this feature, OLM will fallback on existing behavior when checking if an operator can be upgraded.
+
+- Operator may opt-in to this feature but never update the endpoint as "upgradeable". In these situations, cluster admins must uninstall the operator and upgrade the operator manually.
+
+- If OLM reports metrics to Telemeter regarding whether or not an operator is upgradeable an unbounded metric is being reported. An unbounded metric may overwhelm Telemeter.
+
+## Design Details
+
+### Test Plan
+
+OLM's e2e testing suite will be upgraded to handle the usecase:
+
+- If a CSV defines a upgradeReadiness Service without a spec that does not exist, upgrades for that operator are blocked.
+- If a CSVless bundle defines a upgradeReadiness Service that does not exist, the bundle fails to build.
+- If a CSV defines an upgradeReainess Service with a spec, changes to the service are reverted by OLM.
+- If a CSV/CSVless bundle defines an upgradeReainess Service that returns a 0, upgrades to the operator are blocked. The operator then upgrades when the service is no longer blocked.
+
+Additionally, tests that use CSVs that omit the upgradeReainess Service must continue to adhear to existing behavior.
+
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Alternatives
+
+The are viable alternatives to the solution proposed in this enhancement.
+
+### Subsbitute the UpgradeReadinessService with an UpdateProbe
```

Comment:
Would the amount of code/YAML a user would have to write for the `Service` approach defined above be more/less/the same as a probe approach? It seems like probes are easy to define, and I'm not familiar with writing application code exposed via a `Service` endpoint (seems like it would be a simple bit of server code).

Reply:
Total agreement here - After the initial draft, I looked more into the probe approach and believe that it's the best route - pending no opposition/convincing arguments from reviewers.

CC @ecordell @kevinrizza @njhale

Reply:
The previous comment no longer applies in the latest version of this PR after looking into this more.